source/credentials_provider_ecs.c

```diff
@@ -0,0 +1,636 @@
+
+    if (access_key_id_cursor.len == 0 || secret_access_key_cursor.len == 0 || session_token_cursor.len == 0) {
+        goto done;
+    }
+
+    credentials = aws_credentials_new_from_cursors(
+        allocator, &access_key_id_cursor, &secret_access_key_cursor, &session_token_cursor);
+
+    if (creds_expiration_cursor.len != 0) {
+        struct aws_date_time expiration;
+        aws_date_time_init_from_str_cursor(&expiration, &creds_expiration_cursor, AWS_DATE_FORMAT_ISO_8601);
+        credentials->expiration_timepoint_seconds = expiration.timestamp;
+    }
+done:
+
+    if (document_root != NULL) {
+        cJSON_Delete(document_root);
+    }
+
+    return credentials;
+}
+
+/*
+ * No matter the result, this always gets called assuming that esc_user_data is successfully allocated
+ */
+static void s_ecs_finalize_get_credentials_query(struct aws_credentials_provider_ecs_user_data *ecs_user_data) {
+    /* Try to build credentials from whatever, if anything, was in the result */
+    struct aws_credentials *credentials =
+        s_parse_credentials_from_ecs_document(ecs_user_data->allocator, &ecs_user_data->current_result);
+
+    /* pass the credentials back */
+    ecs_user_data->original_callback(credentials, ecs_user_data->original_user_data);
+    if (credentials != NULL) {
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) ECS credentials provider successfully queried instance role credentials",
+            (void *)ecs_user_data->ecs_provider);
+    } else {
+        AWS_LOGF_WARN(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) ECS credentials provider failed to query instance role credentials",
+            (void *)ecs_user_data->ecs_provider);
+    }
+
+    /* clean up */
+    struct aws_credentials_provider *provider = ecs_user_data->ecs_provider;
+    s_aws_credentials_provider_ecs_user_data_destroy(ecs_user_data);
+    aws_credentials_provider_release(provider);
+    aws_credentials_destroy(credentials);
+}
+
+static int s_ecs_on_incoming_body_fn(
+    struct aws_http_stream *stream,
+    const struct aws_byte_cursor *data,
+    void *user_data) {
+
+    (void)stream;
+
+    struct aws_credentials_provider_ecs_user_data *ecs_user_data = user_data;
+    struct aws_credentials_provider_ecs_impl *impl = ecs_user_data->ecs_provider->impl;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p) ECS credentials provider received %zu response bytes",
+        (void *)ecs_user_data->ecs_provider,
+        data->len);
+
+    if (data->len + ecs_user_data->current_result.len > ECS_RESPONSE_SIZE_LIMIT) {
+        impl->function_table->aws_http_connection_close(ecs_user_data->connection);
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) ECS credentials provider query response exceeded maximum allowed length",
+            (void *)ecs_user_data->ecs_provider);
+
+        return AWS_OP_ERR;
```

Comment:
need to aws_raise_error()

Reply:
What kind of error shall we raise is appropriate?

Reply:
Seems like we'll need some new errors, maybe something like `AWS_AUTH_CREDENTIALS_PARSE_ERROR` ?

Reply:
What about AWS_ERROR_SHORT_BUFFER?

Reply:
I think AWS_ERROR_SHORT_BUFFER would be much more confusing to the user than something like PARSE_ERROR